### ISSUE TITLE
[AppConfig] Expired auth token used for retries after 429 error

### ISSUE THREAD

**Issue by avanigupta** [Azure/azure-sdk-for-python#17316]
- **Package Name**: azure-appconfiguration
- **Package Version**: 1.1.1
- **Operating System**: Windows-10-10.0.19041-SP0
- **Python Version**: 3.6.8 

**Describe the bug**
When AppConfig service throws `429` - `Too many requests` error, the data plane SDK retries the request appropriately. However, the same auth token is re-used for every retry. As a result, if a retry is attempted after 15 mins have passed from the original request, AppConfig service will return `401` error because the auth token has expired.

AppConfig service does not accept HMAC auth token if it was issued more than 15 minutes before the current UTC time and throws `401` error as described [here.](https://docs.microsoft.com/en-us/azure/azure-app-configuration/rest-api-authentication-hmac#errors)

The `Date` or `x-ms-date` request header needs to be updated if 401 is received after a non-401 error (or during any retries).

**To Reproduce**

This issue was discovered while stress testing bulk import of key-values from YAML file to App Config store using Azure CLI. From the CLI debug logs and service logs, we can see that `401` error was received after 3 retries and each retry was made 5 mins after the previous retry.

CLI Logs:

```
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16980?api-version=1.0 HTTP/1.1" 429 None
DEBUG: urllib3.connectionpool: Resetting dropped connection: cleanup-test-df1.azconfig-test.io
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16980?api-version=1.0 HTTP/1.1" 429 None
DEBUG: urllib3.connectionpool: Resetting dropped connection: cleanup-test-df1.azconfig-test.io
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16980?api-version=1.0 HTTP/1.1" 429 None
DEBUG: urllib3.connectionpool: Resetting dropped connection: cleanup-test-df1.azconfig-test.io
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16980?api-version=1.0 HTTP/1.1" 401 0
WARNING: cli.azure.cli.command_modules.appconfig._kv_helpers: Failed to set key-value with key 'Key16980' and label 'None'. Operation returned an invalid status 'Unauthorized'
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16981?api-version=1.0 HTTP/1.1" 429 None
DEBUG: urllib3.connectionpool: Resetting dropped connection: cleanup-test-df1.azconfig-test.io
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16981?api-version=1.0 HTTP/1.1" 429 None
DEBUG: urllib3.connectionpool: Resetting dropped connection: cleanup-test-df1.azconfig-test.io
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16981?api-version=1.0 HTTP/1.1" 429 None
DEBUG: urllib3.connectionpool: Resetting dropped connection: cleanup-test-df1.azconfig-test.io
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16981?api-version=1.0 HTTP/1.1" 401 0
WARNING: cli.azure.cli.command_modules.appconfig._kv_helpers: Failed to set key-value with key 'Key16981' and label 'None'. Operation returned an invalid status 'Unauthorized'
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16982?api-version=1.0 HTTP/1.1" 429 None
DEBUG: urllib3.connectionpool: Resetting dropped connection: cleanup-test-df1.azconfig-test.io
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16982?api-version=1.0 HTTP/1.1" 429 None
DEBUG: urllib3.connectionpool: Resetting dropped connection: cleanup-test-df1.azconfig-test.io
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16982?api-version=1.0 HTTP/1.1" 429 None
DEBUG: urllib3.connectionpool: Resetting dropped connection: cleanup-test-df1.azconfig-test.io
DEBUG: urllib3.connectionpool: https://cleanup-test-df1.azconfig-test.io:443 "PUT /kv/Key16982?api-version=1.0 HTTP/1.1" 401 0
WARNING: cli.azure.cli.command_modules.appconfig._kv_helpers: Failed to set key-value with key 'Key16982' and label 'None'. Operation returned an invalid status 'Unauthorized'

```

Service Logs: [https://jarvis-west.dc.ad.msft.net/A0D7256](https://jarvis-west.dc.ad.msft.net/A0D7256)

**Expected behavior**
This can be solved by either regenerating the auth token on every retry, or regenerating the auth token on receiving a `401` during retries.

**Screenshots**

**Additional context**

**Comment by xiangyan99** [comment]
We need to put credential_policy after retry policy

**Comment by seankane-msft** [comment]
This was solved with #17347, @avanigupta if this doesn't solve your issue, please re-open the issue.
